src/mcts/search.cc

```diff
@@ -1291,9 +1307,38 @@ void SearchWorker::ExtendNode(Node* node) {
       return;
     }
 
-    if (history_.Last().GetRepetitions() >= 2) {
+    const auto repetitions = history_.Last().GetRepetitions();
+    const auto twofolddrawlevel = params_.GetTwoFoldDrawLevel();
+    bool validtwofold = false;
+    // Mark two-fold repetitions as draws according to settings
+    // Depth starts with 1 at root, so number of plies in PV is depth - 1
+    if (repetitions >= 2) {
       node->MakeTerminal(GameResult::DRAW);
       return;
+    } else if (repetitions == 1 && twofolddrawlevel > 0) {
+      if (twofolddrawlevel == 3) {
+        // always mark as draw
+        validtwofold = true;
+      } else if (twofolddrawlevel == 2 && depth - 1 >= 4) {
+        // only mark as draw if depth of extended node is >= 4
+        validtwofold = true;
+      } else if (twofolddrawlevel == 1 && depth - 1 >= 4 && depth - 1 >=
+                 history_.ComputePliesSinceFirstRepetition()) {
+        // check whether first repetition happened at root or in the tree
+        // don't mark as draw if repetition happened in the game history
+        validtwofold = true;
+      }
+    }
+    // if node turned out to be a valid twofold, mark it as Terminal::Twofold
+    if (validtwofold) {
```

Comment:
You could just combine the conditions above:
```
if (rep == 1 && level > 0 &&
    (level == 1 && depth …) ||
    (level == 2 && depth >= 3)) {
  MakeTerminal();
  return;
}
```

Reply:
Yes, very sensible suggestion. The current code organization is still from when I thought things were more complicated, I will reorganize it.